extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/user/KsmDirectoryService.java

```diff
@@ -222,6 +231,37 @@ public void processAttributes(Attributes entity) throws GuacamoleException {
 
     }
 
+    @Override
+    public Directory<Connection> getConnectionDirectory(
+            Directory<Connection> underlyingDirectory) throws GuacamoleException {
+
+        // A Connection directory that will decorate all connections with a
+        // KsmConnection wrapper to ensure that all defined KSM fields will
+        // be exposed in the connection group attributes.
+        return new DecoratingDirectory<Connection>(underlyingDirectory) {
+
+            @Override
+            protected Connection decorate(Connection connection) throws GuacamoleException {
+
+                // Wrap in a KsmConnection class to ensure that all defined KSM fields will be
+                // present
+                return new KsmConnection(
+                        connection,
+                        ksmAttributeService.getConnectionAttributes().stream().flatMap(
+                                form -> form.getFields().stream().map(field -> field.getName())
+                        ).collect(Collectors.toList()));
```

Comment:
Feels somewhat wonky to pull this dynamically when we have `KSM_CONNECTION_ATTRIBUTES` available statically and the stated purpose of `KsmConnection` is to ensure those attributes are defined. Why not have `KsmConnection` leverage `KSM_CONNECTION_ATTRIBUTES` internally such that it need not be fed each attribute name manually?

Reply:
Yeah, that would make more sense.